CenterForOpenScience/ember-osf-web
# [ENG-1641] Draft registration save before exit

-   Ticket: [ENG-1641]
-   Feature flag: n/a

## Purpose
- Let users know they have unsaved changes before they exit the draft registration workflow
- Let users know when an autosave is in progress

## Summary of Changes
- Change some of the links on the Draft Registration Metadata page to open in a new tab to prevent users from navigating away
- Add a new `{{before-unload}}` modifier
- Add logic to show popup if the draft has pending changes
  - One to handle `onBeforeUnload` (closing or refreshing the browser window)
  - One to handle route transitions (clicking the "Add new"/"My Registrations" buttons or navigating away within the app)
- Add same logic for edit-revision workflow
- Update rightnav template to show when an autosave is in progress

## Screenshot(s)
- Edit revision workflow. When navigating away within 3 seconds of updating the revision justification:
![image](https://user-images.githubusercontent.com/51409893/189439437-f408a409-5c96-4349-bf86-cea612b65725.png)

- Draft registration workflow. When trying to refresh the tab within 3 seconds of updating a response:
![image](https://user-images.githubusercontent.com/51409893/189439732-653b62ca-fa37-4abc-8546-d042a6dce657.png)

- Autosave indicator:
![image](https://user-images.githubusercontent.com/51409893/190270478-61150bfd-3b7d-4f0e-ab70-552ba01071cd.png)

## Side Effects
- I've made it so that clicking the contributor names link, as the license help link, on the Metadata page opens in a new tab. Less of a side-effect, but more a preference.

## QA Notes
- There is a 3 second delay before the last user input and when the autosave kicks off. If a user tries to close the tab or reload the page during that 3 second window, a prompt should show up asking if they _really_ want to leave. If they hit cancel, the save should trigger and once successful or unsuccessful, a toast message should appear indicating the save went through or not


[ENG-1641]: https://openscience.atlassian.net/browse/ENG-1641?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
